django/middleware/csrf.py

```diff
@@ -154,15 +157,35 @@ def process_view(self, request, callback, callback_args, callback_kwargs):
                 if referer is None:
                     return self._reject(request, REASON_NO_REFERER)
 
+                referer = urlparse(referer)
+
+                # Make sure we have a valid URL for Referer.
+                if '' in (referer.scheme, referer.netloc):
+                    return self._reject(request, REASON_MALFORMED_REFERER)
+
+                # Ensure that our Referer is also secure.
+                if referer.scheme != 'https':
+                    return self._reject(request, REASON_INSECURE_REFERER)
```

Comment:
I think rejecting an insecure referer on secure requests is an existing behavior. Here's a test for Django 1.8.x:
```
@override_settings(DEBUG=True)
def test_https_reject_insecure_referer(self):
    """
    A POST HTTPS request from an insecure referer should be rejected.
    """
    req = self._get_POST_request_with_token()
    req._is_secure_override = True
    req.META['HTTP_REFERER'] = 'http://example.com/'
    req.META['SERVER_PORT'] = '443'
    req.META['SERVER_NAME'] = 'example.com'
    response = CsrfV
```
The response says "Referer checking failed - http://example.com/ does not match https://example.com/"